src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -340,13 +850,148 @@ internal static unsafe float SumAbsU(Span<float> src)
+                    srcVector = Sse.SubtractScalar(srcVector, meanVector);
+                    result = Sse.AddScalar(result, Sse.And(srcVector, mask));
+
+                    pSrcCurrent++;
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static unsafe float MaxAbsU(Span<float> src)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+            Vector128<float> mask;
+
+            if (Sse2.IsSupported)
+            {
+                mask = Sse.StaticCast<int, float>(Sse2.SetAllVector128(0x7FFFFFFF));
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(BitConverter.Int32BitsToSingle(0x7FFFFFFF));
+            }
+
+            fixed (float* psrc = src)
+            {
+                float* pSrcCurrent = psrc;
+                float* pEnd = psrc + src.Length;
+
+                while (pSrcCurrent + 4 <= pEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                    result = Sse.Max(result, Sse.And(srcVector, mask));
+
+                    pSrcCurrent += 4;
+                }
+
+                Vector128<float> x1 = Sse.Shuffle(result, result, 0xB1);
+                result = Sse.Max(result, x1);
+                x1 = Sse.Shuffle(result, result, 0x02);
+                result = Sse.MaxScalar(result, x1);
+
+                while (pSrcCurrent < pEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                    result = Sse.MaxScalar(result, Sse.And(srcVector, mask));
+
+                    pSrcCurrent++;
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static unsafe float MaxAbsDiffU(float mean, Span<float> src)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+            Vector128<float> meanVector = Sse.SetAllVector128(mean);
+            Vector128<float> mask;
+
+            if (Sse2.IsSupported)
+            {
+                mask = Sse.StaticCast<int, float>(Sse2.SetAllVector128(0x7FFFFFFF));
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(BitConverter.Int32BitsToSingle(0x7FFFFFFF));
+            }
+
+            fixed (float* psrc = src)
+            {
+                float* pSrcCurrent = psrc;
+                float* pEnd = psrc + src.Length;
+
+                while (pSrcCurrent + 4 <= pEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                    srcVector = Sse.Subtract(srcVector, meanVector);
+                    result = Sse.Max(result, Sse.And(srcVector, mask));
+
+                    pSrcCurrent += 4;
+                }
+
+                Vector128<float> x1 = Sse.Shuffle(result, result, 0xB1);
```

Comment:
These 4 lines are duplicated with the other Max method.  Can we extract a common method here?

Reply:
Agree and thanks!

Should I make `SumAbsU` call `SumAbsDiffU` and `MaxAbsU` call `MaxAbsDiffU` with `mean == 0f`?

Reply:
No, I wouldn't do it that way.  If you did it that way, you always would need to emit `Subtract(0)` calls, which would slow the function down.

What I mean is just these 4 duplicated lines that take the `result`, and get the MaxScalar value from it.  Similar to the `VectorSum` method.